test/hotspot/jtreg/compiler/c2/irTests/TestScopedValue.java

```diff
@@ -0,0 +1,708 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.irTests;
+
+import compiler.lib.ir_framework.*;
+import compiler.lib.ir_framework.test.TestVM;
+import jdk.test.whitebox.WhiteBox;
+import java.lang.reflect.Method;
+import compiler.whitebox.CompilerWhiteBoxTest;
+import jdk.test.lib.Platform;
+
+import java.util.Arrays;
+import java.util.List;
+
+/*
+ * @test
+ * @bug 8320649
+ * @summary C2: Optimize scoped values
+ * @library /test/lib /
+ * @build jdk.test.whitebox.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
+ * @compile --enable-preview -source ${jdk.version} TestScopedValue.java
+ * @run main/othervm --enable-preview -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI compiler.c2.irTests.TestScopedValue
+ */
+
+public class TestScopedValue {
+
+    private static final WhiteBox WHITE_BOX = WhiteBox.getWhiteBox();
+
+    private static long tieredStopAtLevel = (long)WHITE_BOX.getVMFlag("TieredStopAtLevel");
+
+    static ScopedValue<MyDouble> sv = ScopedValue.newInstance();
+    static final ScopedValue<MyDouble> svFinal = ScopedValue.newInstance();
+    static ScopedValue<Object> svObject = ScopedValue.newInstance();
+    private static volatile int volatileField;
+
+    public static void main(String[] args) {
+        if (Platform.isComp()) {
+            TestFramework.runWithFlags("--enable-preview");
+        } else {
+            // Fast path tests need to be run one at a time to prevent profile pollution
+            List<String> tests = List.of("testFastPath1", "testFastPath2", "testFastPath3", "testFastPath4",
+                    "testFastPath5", "testFastPath6", "testFastPath7", "testFastPath8", "testFastPath9",
+                    "testFastPath10", "testFastPath11", "testFastPath12", "testFastPath13", "testFastPath14", "testFastPath15",
+                    "testSlowPath1,testSlowPath2,testSlowPath3,testSlowPath4,testSlowPath5,testSlowPath6,testSlowPath7,testSlowPath8,testSlowPath9,testSlowPath10");
+            for (String test : tests) {
+                TestFramework.runWithFlags("-XX:+TieredCompilation", "--enable-preview", "-XX:CompileCommand=dontinline,java.lang.ScopedValue::slowGet", "-DTest=" + test);
```

Comment:
What is the reason for running each test individually?

Reply:
Hmm. Profile pollution. But if it is so bad, then won't that be an issue "in the real wold"? Is this test not very artificial?

Reply:
What I would expect is that if there are only a few uses of ScopedValue, then there's a good chance there's no profile pollution. If there are many and profile is polluted, this PR implements some optimizations that should do a reasonable job.